design-proposals/klip-4-custom-types.md

```diff
@@ -0,0 +1,190 @@
+  # KLIP-4: Custom Type Registry
+
+**Author**: agavra | 
+**Release Target**: 5.4 | 
+**Status**: In Discussion | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/2894
+
+**tl;dr:** *Introduce a feature that makes custom types easier to work with in KSQL by aliasing
+complex type declarations.*
+
+## Motivation and background
+
+Kafka often contains very complex nested data structures and it is common to see schemas nested with
+other schemas. It is desirable to be able to define this custom schema in KSQL once and reference it
+in all future statements. For example, imagine the following schema declarations:
+
+```sql
+CREATE TYPE ADDRESS AS STRUCT<number INTEGER, street VARCHAR, city VARCHAR>;
+CREATE TYPE PERSON AS STRUCT<firstname VARCHAR, lastname VARCHAR, address ADDRESS>;
+CREATE TYPE COMPANY AS STRUCT<name VARCHAR, headquarters ADDRESS>;
+```
+
+With this feature organizations can leverage existing schemas in their KSQL queries and share 
+data with other parts of their organizations.
+
+## Scope
+
+* Syntax to register and unregister custom type aliases
+* Ability to include custom types at compile-time (e.g. in the JAR) for use in UDF declarations
+* Ability to list all custom types
+
+Schema Registry integration is not considered within scope.
+
+## Value/Return
+
+Real data is complex, nested and referential. This feature will help us dramatically improve the
+user interaction with complicated schemas. 
+
+## Public APIS
+
+### CREATE TYPE
+
+```sql
+CREATE TYPE <type_name> AS <type>;
+```
+
+The `CREATE TYPE` syntax will allow KSQL users to register a type alias directly in SQL (either 
+interactive or headless modes). Any types registered using this command can be leveraged in any
+future statement. 
+
+Any attempts to register the same type twice without a corresponding `DROP TYPE` will fail.
```

Comment:
I assume we'll also return an error if the user tries registering (or dropping) a type that conflicts with an existing type name, such as `DOUBLE`. What about other reserved KSQL words that aren't types, such as `TABLE`?

Reply:
Sure :) I'll add in those restrictions since we can always left it later